src/aggregated_bls/party_i.rs

```diff
@@ -52,11 +57,41 @@ impl Keys {
 
     pub fn combine_local_signatures(sigs: &[SIG]) -> BLSSignature {
         let (head, tail) = sigs.split_at(1);
-        let sig_sum = tail.iter().fold(head[0], |x, acc| x + acc);
+        let sig_sum = tail.iter().fold(head[0], |acc, x| acc + x);
         BLSSignature { sigma: sig_sum }
     }
 
     pub fn verify(signature: &BLSSignature, message: &[u8], apk: &APK) -> bool {
         signature.verify(message, apk)
     }
+
+    fn efficient_pairing_loop(vec_g1: &[GE1], vec_g2: &[GE2]) -> Pair {
+        let vec_g1_prep: Vec<_> = vec_g1.iter().map(|x| x.get_element().prepare()).collect();
+        let vec_g2_prep: Vec<_> = vec_g2.iter().map(|x| x.get_element().prepare()).collect();
+        let vec: Vec<_> = vec_g1_prep.iter().zip(vec_g2_prep.iter()).collect();
+        Pair {
+            e: Bls12::final_exponentiation(&Bls12::miller_loop(vec.iter())).unwrap(),
+        }
+    }
+
+    pub fn aggregate_bls(sig_vec: &[BLSSignature]) -> BLSSignature {
+        let (head, tail) = sig_vec.split_at(1);
+        BLSSignature {
+            sigma: tail.iter().fold(head[0].sigma, |acc, x| acc + x.sigma),
+        }
+    }
+
+    fn core_aggregate_verify(apk_vec: &[APK], msg_vec: &[&[u8]], sig: &BLSSignature) -> bool {
+        let product_c2 = Keys::efficient_pairing_loop(&[sig.sigma], &[GE2::generator()]);
+        let vec_g1: Vec<GE1> = msg_vec.iter().map(|&x| GE1::hash_to_curve(&x)).collect();
+        let product_c1 = Keys::efficient_pairing_loop(&vec_g1, &apk_vec);
+        product_c1.e == product_c2.e
+    }
+
+    pub fn aggregate_verify(apk_vec: &[APK], msg_vec: &[&[u8]], sig: &BLSSignature) -> bool {
```

Comment:
according to paper : here we need to check that all messages are different. 
If some messages are the same : we need to use randomly selected exponents. see papers

Reply:
The [BLS signatures draft](https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-04#section-3.1.1) simply returns invalid when there are repeated messages. There is also the option to implement the [message augmentation scheme](https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-04#section-3.2). I stuck to the draft, but we could implement any of the three.